### ISSUE TITLE
AArch64 cannot select `init_trampoline`

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#27174]
The good news is that the AArch64 builds are back, the bad new is that the test are failing.
From https://build.julialang.org/#/builders/86/builds/206/steps/2/logs/stdio

```
      From worker 7:	LLVM ERROR: Cannot select: 0x3b0167f8: ch = init_trampoline 0x30660570, 0x34ac35e0, 0x3b2fd438, 0x3af9f838, SrcValue:ch<0x3b333810>, SrcValue:ch<0x3b194f08>
      From worker 7:	  0x34ac35e0: i64,ch = CopyFromReg 0x30660570, Register:i64 %0
      From worker 7:	    0x3b2fd980: i64 = Register %0
      From worker 7:	  0x3b2fd438: i64 = AArch64ISD::WrapperLarge TargetGlobalAddress:i64<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=3], TargetGlobalAddress:i64<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=36], TargetGlobalAddress:i64<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=37], TargetGlobalAddress:i64<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=38]
      From worker 7:	    0x3b0187b0: i64 = TargetGlobalAddress<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=3]
      From worker 7:	    0x34b21630: i64 = TargetGlobalAddress<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=36]
      From worker 7:	    0x3b2fd7e0: i64 = TargetGlobalAddress<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=37]
      From worker 7:	    0x3b016e78: i64 = TargetGlobalAddress<%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)**, i8*)* @jlcapi_cfunction_38022> 0 [TF=38]
      From worker 7:	  0x3af9f838: i64,ch = CopyFromReg 0x30660570, Register:i64 %1
      From worker 7:	    0x3b018678: i64 = Register %1
      From worker 7:	In function: jlcapi_cfunction_38022make
```

@vtjnash is this the issue with trampoline suppport on ARM that you mentioned a while back?

**Comment by vtjnash** [comment]
Yeah, we need to edit our lowering to emit an explicit “not yet implemented on ARM”

**Comment by tianon** [comment]
Is this why 0.6.4 lost the pre-compiled armv7l and aarch64 release artifacts? :cry:

(Can't seem to find an issue/PR/change referencing the removal of those two tarballs for 0.6.4.)

**Comment by oschulz** [comment]
Just stumbled on this via FunctionWrappers.jl - is this still an issue? Just wondering since it may affect the design of a project of ours.

**Comment by vchuravy** [comment]
Yes still an issue on PowerPC and AArch64 afaik

**Comment by oschulz** [comment]
Will this have any practical implications for Apple-Silicon systems?

**Comment by musm** [comment]
We have reports of users running into these problem at https://github.com/JuliaIO/HDF5.jl/issues/811#issuecomment-766424062
Does LLVM 11 improve the situation here at all?

**Comment by vchuravy** [comment]
I just checked. It seems PPC has gained support for trampoline's, but we end up calling `abort` from within the `__init_trampoline` call (https://android.googlesource.com/toolchain/gcc/+/master/gcc-4.9/libgcc/config/rs6000/tramp.S#181).

Grepping for `INIT_TRAMPOLINE` shows that only the X86 and PPC backend have lowerings.

**Comment by giordano** [comment]
If I understand it correctly, this might be addressed upstream by https://github.com/llvm/llvm-project/pull/70267

**Comment by vchuravy** [comment]
The issue with the LLVM implementation is that it uses x18 which is a reserved register on Apple.

GCC 14 has recently added heap-trampolines https://gcc.gnu.org/onlinedocs/gcc-14.1.0/gcc/Code-Gen-Options.html#index-ftrampoline-impl and they use `x16` for the context variable.

**Comment by vtjnash** [comment]
We have only ever used heap trampolines, but we need LLVM to implement it first. In theory it is quite easy, since it can choose any unused register, save it, and then replace it with the trampoline value or choose any offset from the stack pointer address and put the value there. GCC defines quite a few different valid implementation options for the static chain value: https://gcc.gnu.org/onlinedocs/gcc-3.1/gccint/Frame-Registers.html#:~:text=The%20static%20chain%20register%20need,two%20macros%20should%20be%20defined.&text=STATIC_CHAIN%20STATIC_CHAIN_INCOMING-,If%20the%20static%20chain%20is%20passed%20in%20memory%2C%20these%20macros,denote%20where%20they%20are%20stored.

(Most platforms have had a free scratch register too, making it even easier to implement, but that is just an optimization)

**Comment by 3f6a** [comment]
I'm hitting this error:

```
ERROR: LoadError: cfunction: closures are not supported on this platform
```

coming from this line of code:

```
@cfunction($move_to,  Cint, (Ptr{FreeType.FT_Vector}, Ptr{Cvoid}))
```

This code works fine on a Linux machine but fails on a macos (Apple Silicon arm).

**Comment by vtjnash** [comment]
Bug fixed upstream in future llvm versions

**Comment by 3f6a** [comment]
> future llvm versions

You mean this is not yet available in latest Julia (1.12.5) ? We have to wait until Julia updates LLVM

**Comment by vtjnash** [comment]
Yes, though note that if you have a `Ptr{Cvoid}` context argument, there are far more efficient and performant ways of passing in the julia function (see the ccall documentation for examples of qsort_r)
